Build out StringIO support

bobjacobsen opened the pull request:
This builds out the StringIO support.  It includes 
 - a StringIO Table, 
 - help pages with index and TOC updates, 
 - support for opening the StringIO Table at start,
 - an internal implementation,
 - persistence for the internal StringIO objects.
 
It does not include (yet)
 - adding StringIO icons to the various panel types,
 - a general implementation for LocoNet StringIOs (there's a throttle-specific on that doesn't do anything).
 
There's also a little bit of cleanup included.

dsand47 commented on help/en/html/tools/StringIO.shtml:
```diff
@@ -0,0 +1,51 @@
+      </ul>
+
+      <h2>Using StringIOs with Logix</h2>
+      A Logix can test a StringIO for particular contents, and set the contents of a StringIO to
```
This should be LogixNG, not Logix.

bobjacobsen replied:
Thanks!  Fixed by changing to "Logix or LogixNG" there and in the Memories page.

dsand47 commented on xml/schema/layout-5-5-5.xsd:
```diff
@@ -55,6 +55,7 @@ signals - obsolete, EMPTY, never used
   <xs:include schemaLocation="http://jmri.org/xml/schema/types/lights-4-19-2.xsd"/>
   <xs:include schemaLocation="http://jmri.org/xml/schema/types/reporters-2-9-6.xsd"/>
   <xs:include schemaLocation="http://jmri.org/xml/schema/types/memories-2-9-6.xsd"/>
+  <xs:include schemaLocation="http://jmri.org/xml/schema/types/stringios-5-11-2.xsd"/>
```
The referenced file is not included in the PR.

bobjacobsen replied:
Thanks. I've added it.

dsand47 commented on help/en/html/tools/StringIO.shtml:
```diff
@@ -0,0 +1,52 @@
+      </ul>
+
+      <h2>Using StringIOs with Logix</h2>
+      A Logix or LogixNG can test a StringIO for particular contents, 
```
Logix has not been updated to include support for StringIO.

bobjacobsen replied:
Thanks, I didn't think of that. Fixed several places.

dsand47 commented:
Some minor issues.

- When a value is updated, the table display is not updated until the window is made active.
- The LocoNet beans are not stored.  The comments in the xml class indicates that this is deferred.
- The schema has a "value" attribute but it is not being used.  Future feature?
- There are number of places where the Javadoc has prior text due to copy/paste.
- @danielb987 will need to update LogixNG to implement StringIO read and expressions.

Table management works including veto processing when used by a LogixNG.  Store and Load work for Internal beans.

danielb987 commented:
> * @danielb987 will need to update LogixNG to implement StringIO read and expressions.

I will, but I cannot complete it until this PR is merged.

> * The schema has a "value" attribute but it is not being used.  Future feature?

I think it would be good to have the _option_ to store the value in the tables and panels file. That way, you could initialize the StringIO to a default value if desired. I'm thinking of a "Store value" checkbox in the StringIO table and if selected, the value is stored and loaded. For OpenLCB/LCC shared memories, when a StringIO with a stored value is loaded, the loaded value is then set on the LCC bus.
But this feature can be added later if desired.

bobjacobsen commented:
Thanks for the careful checking!

Some minor issues.

> When a value is updated, the table display is not updated until the window is made active.

That's interesting.  Do you use two open StringIO Table windows to see this, or some other method?

> The LocoNet beans are not stored. The comments in the xml class indicates that this is deferred.

Right.  There's something odd about the LocoNet support that I haven't figured out yet.  It seems to automatically create a throttle-specific `StringIO` object each time it's started up.  

> The schema has a "value" attribute but it is not being used. Future feature?

Yes.  I didn't want to have to update the schema if we decide to store and load the current value.  I think Daniel's suggestion to make that user-selectable is a good one.  I'll add it, but not right away.

> There are number of places where the Javadoc has prior text due to copy/paste.

Will try to track those down.

danielb987 commented:
> It does not include (yet)
> 
>     * a general implementation for LocoNet StringIOs (there's a throttle-specific on that doesn't do anything).

The existing LocoNet StringIO is `LCThrottles`. And it's working.

Some hardware LocoNet throttles have a feature where you can send a text message on the LocoNet and that message is then shown on the throttles that supports it. In the `LocoNet` menu in PanelPro, there is the menu item `Send Throttle Messages`. If you open that window and enters a text, that text will be sent to the LocoNet throttles.

The LocoNet StringIO `LCThrottles` has the same feature. If you write a text to that StringIO, the text will be sent to the LocoNet throttles.

This LogixNG demonstrates it:
```
LogixNG: Test StringIO
    ConditionalNG: IQC:AUTO:0004 ::: Startup
        ! A
            Read string E and set string A
                ?s E
                    Get string constant "Hello"
                !s A
                    Set stringIO LCThrottles
```

Open the LocoNet Monitor and select `Show raw data` and you will see that the LogixNG sends the same data as the `Send Throttle Messages` window.

This StringIO is `jmri.jmrix.loconet.LnThrottleStringIO`. There is only one of it. It's created and own by the LocoNetSystemConnectionMemo. The creation is done in the `getThrottleStringIO()` of the memo.

The LocoNet StringIO `LCThrottles` is similar to the Memory IMCURRENTTIME that's always created when JMRI starts. If you have a LocoNet connection, you are able to send the throttle message to LocoNet and therefore the LocoNet StringIO `LCThrottles` is always created when the LocoNet connection is created.

dsand47 commented:
> That's interesting. Do you use two open StringIO Table windows to see this, or some other method?

I used a LogixNG to update the value.  I also updated a memory value at the same time.  The tables were in separate windows.  The memory value updated immediately, but the StringIO value did not update until I clicked on the window frame.

danielb987 commented on java/src/jmri/StringIOManager.java:
```diff
@@ -1,10 +1,54 @@
+     * access of the desired sensor, and the user address is associated with it.
+     * The system name must be valid.
+     * </ul>
+     * Note that it is possible to make an inconsistent request if both
+     * addresses are provided, but the given values are associated with
+     * different objects. This is a problem, and we don't have a good solution
+     * except to issue warnings. This will mostly happen if you're creating
+     * StringIOs when you should be looking them up.
```
Since this method can throw an `lllegalArgumentException`, we actually have a very good way to handle it. Throw an `lllegalArgumentException` if the system name and user name refers to different beans.

bobjacobsen replied:
That's what it does.

danielb987 commented on java/src/jmri/implementation/DefaultStringIO.java:
```diff
@@ -0,0 +1,50 @@
+        super(systemName, userName);
+    }
+
+    /**
+     * Sends the string to the layout.
+     * The string [u]must not be longer than the value of getMaximumLength()
+     * unless that value is zero. Some microcomputers have little memory and
+     * it's very important that this method is never called with too long strings.
+     *
+     * @param value the desired string value
+     * @throws jmri.JmriException general error when setting the value fails
+     */
```
I suggest removing the comment and replace it with `/** {@inheritDoc} */` instead.

bobjacobsen replied:
Done, thanks.

danielb987 commented on java/src/jmri/jmrit/beantable/StringIOTableAction.java:
```diff
@@ -0,0 +1,319 @@
+            return;
+        }
+        String rName;
+        String stringIOPrefix = prefixBox.getSelectedItem().getSystemPrefix(); // Add "R" later
```
`// Add "R" later`
What does `R` stands for? Copy/paste error?

bobjacobsen replied:
I don't know what this comment means.  it comes from the Reporter support, which I copied to create this.  I'll remove it from StringIO and leave it in Reporter.

danielb987 commented on java/src/jmri/jmrix/loconet/LnStringIOManager.java:
```diff
@@ -49,6 +52,25 @@ public void dispose() {
+    @Override
+    @Nonnull
+    public StringIO provideStringIO(@Nonnull String sName) throws IllegalArgumentException {
+        return new DefaultStringIO(sName);
```
I strongly suggest throwing an `IllegalArgumentException` here for now.

`throw new IllegalArgumentException("Generic LocoNet StringIOs are not yet supported");`

bobjacobsen replied:
> I strongly suggest throwing an IllegalArgumentException here for now.

I changed it three places where the LnStringIoManager was created a StringIO

danielb987 commented on java/src/jmri/managers/AbstractStringIOManager.java:
```diff
@@ -32,6 +34,67 @@ public int getXMLOrder() {
+            s = getByUserName(userName);
+            if (s != null) {
+                if (getBySystemName(systemName) != s) {
+                    log.error("inconsistent user ({}) and system name ({}) results; userName related to ({})", userName, systemName, s.getSystemName());
```
I suggest throwing an `IllegalArgumentException` here.

bobjacobsen replied:
> I suggest throwing an IllegalArgumentException here.

This is probably a good idea, but this is what all of these managers do now: They return the existing object with the requested name, and issue that error.  We can change that behavior here, but I'm not sure I want this manager to behave like every other one in the system.  And changing all those other ones may well break scripts, etc where the user hasn't noticed any problems.

danielb987 commented on java/src/jmri/managers/AbstractStringIOManager.java:
```diff
@@ -32,6 +34,67 @@ public int getXMLOrder() {
+            if ((s.getUserName() == null) && (userName != null)) {
+                s.setUserName(userName);
+            } else if (userName != null) {
+                log.warn("Found StringIO via system name ({}) with non-null user name ({}). StringIO \"{}({})\" cannot be used.",
+                        systemName, s.getUserName(), systemName, userName);
```
I suggest throwing an `IllegalArgumentException` here.

bobjacobsen replied:
This is the counterpart of the suggestion above. This is probably a good idea, but this is what all of these managers do now: They return the existing object with the requested name, and issue that error.  We can change that behavior here, but I'm not sure I want this manager to behave like every other one in the system.  And changing all those other ones may well break scripts, etc where the user hasn't noticed any problems.  Are we sure we want to change just this one?

And thank you for the careful reviews!

danielb987 replied:
I thought that it would be good to change all the managers but I realize that it could cause problems with scripts.
I'm fine with keeping it as is.

danielb987 reviewed (commented):
Thank you!

bobjacobsen commented on java/src/jmri/NamedBeanBundle.properties:
```diff
@@ -458,6 +459,7 @@ SendToolTip                      = Send the specified packet (message)
 EnterHexToolTip                  = Enter command as hex string (0xnnn)
 EnterHexBytesToolTip             = Enter command as hexadecimal bytes separated by a space, for example: 82 7D
 EnterNumber1to9999ToolTip        = Enter a number from 1 to 9999
+EnterStringToolTip               = Enter a string without $
```
Misremembered.  Is should actually be without $, : or ".  See https://www.jmri.org/help/en/html/doc/Technical/Names.shtml
Fixed.

bobjacobsen commented on java/src/jmri/StringIOManager.java:
```diff
@@ -1,10 +1,54 @@
+     * <p>
+     * Note that
+     * two calls with the same arguments will get the same instance; there is
+     * only one StringIO object representing a given physical turnout and
```
> turnout is a bit misleading here.
Fixed

> But a question: If I call newStringIO("Daniel") and the StringIO LNDaniel exists, then the StringIO LNDaniel will be returned?

The search order is user name exactly, system name exactly, then in the default system with the prefix added. So if there's an LNDaniel in the LocoNet (default) system, it'll be found.  Otherwise, not.

bobjacobsen commented on java/src/jmri/StringIOManager.java:
```diff
@@ -1,10 +1,54 @@
+     * only one StringIO object representing a given physical turnout and
+     * therefore only one with a specific system or user name.
+     * <p>
+     * This will always return a valid object reference; a new object will be
+     * created if necessary. In that case:
```
In that case, it didn't return.  Exceptions don't have a return value.  I'll clarify the comment in this case, but it's common across all our managers.

danielb987 commented:
PR looks good to me. I will wait for @dsand47 to give his view on the PR too.

> It does not include (yet)
> 
>     * adding StringIO icons to the various panel types,

See  #11647. It adds icons for LogixNG Global Variables to Panel Editor and Layout Editor. A Global Variable is similar to Memory and StringIO.

dsand47 commented:
Jython access has an error.

The program:

```
import java
import jmri

print 'SIO = {}'.format(stringios)

for sio in stringios.getNamedBeanSet():
    print sio.getUserName()

nm = stringios.getStringIO('newname')
print nm
```

The getStringIO fails.

```
09:58:50,226 jmri.configurexml.LoadXmlConfigAction INFO  - Loading selected file: /Users/das/JMRI/_Profiles/LogixNG.jmri/stringio.xml [AWT-EventQueue-0]
SIO = jmri.managers.ProxyStringIOManager@4bc3c7ce
newname
None
09:59:02,143 jmri.jmrit.jython.RunJythonScript     ERROR - Unable to execute script. [AWT-EventQueue-0]
javax.script.ScriptException: AttributeError: 'jmri.managers.ProxyStringIOManager' object has no attribute 'getStringIO' in <script> at line number 9
```

danielb987 commented:
Dave,

There is no method getStringIO() in the manager. I suppose you thought about newStringIO(). You can also use the method provide().

Daniel

26 dec. 2024 17:10:33 Dave Sand ***@***.***>:

> 
> Jython access has an error.
> 
> The program:
> 
> *import java
> import jmri
> 
> print 'SIO = {}'.format(stringios)
> 
> for sio in stringios.getNamedBeanSet():
>     print sio.getUserName()
> 
> nm = stringios.getStringIO('newname')
> print nm
> *
> The getStringIO fails.
> 
> *09:58:50,226 jmri.configurexml.LoadXmlConfigAction INFO  - Loading selected file: /Users/das/JMRI/_Profiles/LogixNG.jmri/stringio.xml [AWT-EventQueue-0]
> SIO = ***@***.***
> newname
> None
> 09:59:02,143 jmri.jmrit.jython.RunJythonScript     ERROR - Unable to execute script. [AWT-EventQueue-0]
> javax.script.ScriptException: AttributeError: 'jmri.managers.ProxyStringIOManager' object has no attribute 'getStringIO' in <script> at line number 9
> *
> —
> Reply to this email directly, view it on GitHub[https://github.com/JMRI/JMRI/pull/13710#issuecomment-2562919937], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AE2REVNUZ3XFCOGFEJ7DFED2HQS7NAVCNFSM6AAAAABUFATJ4GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNRSHEYTSOJTG4].
> You are receiving this because you were mentioned.
> [Spårningsbild][https://github.com/notifications/beacon/AE2REVP2Q62PFA7VNJDT7ZL2HQS7NA5CNFSM6AAAAABUFATJ4GWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTUYYMHAC.gif]
>

dsand47 commented:
> Dave, There is no method getStringIO() in the manager. I suppose you thought about newStringIO(). You can also use the method provide(). Daniel 26 dec. 2024 17:10:33 

It needs to be added to the StringIOManager interface and subsequent classes to be consistent with other beans.

bobjacobsen commented:
Thanks for the test script.  I've added `getStringIO` to the interface and implementation, script runs fine.  Also fixed a `Turnout` reference I found in the Memory Javadoc, and added the start of a release note.

dsand47 commented:
I have extended the script to change values.
```
import java
import jmri

print 'SIO = {}'.format(stringios)

for sio in stringios.getNamedBeanSet():
    print sio.getUserName()

nm = stringios.getStringIO('newname')
print nm

if nm is not None:
    print 'curr value = {}'.format(nm.getCommandedStringValue())
    nm.setCommandedStringValue('xyz')
    print 'c: new value = {}'.format(nm.getCommandedStringValue())
    print 'k: new value = {}'.format(nm.getKnownStringValue())
```

It runs without errors.
```
SIO = jmri.managers.ProxyStringIOManager@7e78e6a1
newname
None
IC123
curr value = qrs
c: new value = xyz
k: new value = 
```

I am not clear about commanded vs. known.

bobjacobsen commented:
> I am not clear about commanded vs. known.

Neither am I.  it may be by analogy to Turnouts, but I don't see what would create meaningful feedback for StringIOs.

@danielb987 what did you have in mind here?

bobjacobsen commented:
@danielb987 If we changed the `StringIO` interface to just have `String getValue()` and `void setValue(String ..)`, would that badly break anything in LogixNG?  That's similar to the interface that `Memory` objects have, and it might be more straight-forward for people to understand.

danielb987 commented:
> @danielb987 If we changed the `StringIO` interface to just have `String getValue()` and `void setValue(String ..)`, would that badly break anything in LogixNG? That's similar to the interface that `Memory` objects have, and it might be more straight-forward for people to understand.

I'm fine with that.

I added StringIO and AnalogIO at the same time. For StringIO, there is probably no need for Commanded Value. But for AnalogIO, there is.

An example is a crane with three servos. Each servo is an AnalogIO (an integer value 0 <= x <= 270 degrees). The commanded value tells where you want the servo to go. The known value tells where the servo currently is.

bobjacobsen commented:
I'm doing a little more work before deciding on "commanded and known values" vs "just one value".  There are arguments both ways, so I'm doing some work on the OpenLCB implementation to see which one makes more sense.  Might be a day or so before I'm back with more info.

danielb987 commented:
I vote for keeping "commanded and known values". Memory is internal only so it doesn't make any sense to have it for Memory. But StringIO is mainly for the layout so there it makes sense.

A use case is if you have an AnalogIO on top of a StringIO. For example a LocoNet device that's using a couple of AnalogIOs to control the servos of a crane. The AnalogIO converts the integer value to/from a string and then uses a StringIO internally to send the data to and from the LocoNet device. Since LNSV2 on LocoNet can both read and write data to a device, it makes sense that the StringIO have both commanded value (where the servo should go) and known value (where the servo currently is).

danielb987 commented:
@bobjacobsen 
There is a piece missing in StringIO that I have thought about adding after this PR is merged. But you might be interested in adding it now. You mentioned that OpenLCB/LCC shared memory is a byte stream. I think it would be very useful if StringIO had the option to internally use a byte stream to communicate with the layout.

The reason I have a method `canSetCharSet()` is if the hardware is able to tell JMRI which charset the hardware uses, for example UTF8. In that case, there is no point of letting the user change the charset of the StringIO.

For an internal StringIO, this conversion can be ignored. There is no point of doing the conversion for DefaultStringIO.

```
StringIO interface

    /**
     * Set the charset.
     * throws UnsupportedOperationException if this StringIO cannot change the charset.
     *         This exception must not be thrown unless canSetCharSet() returns false.
     */
    void setCharSet(@Nonnull CharSet charset) throws UnsupportedOperationException;

    /**
     * Get the current charset.
     */
    @Nonnull
    CharSet getCharSet();

    /**
     * Does this StringIO supports changing the charset?
     */
    boolean canSetCharSet();

AbstractStringIO

    /**
     * Send byte data to the layout.
     */
    public abstract void sendToLayout(@Nonnull byte[] data);

    /**
     * Receive byte data from the layout.
     */
    public void receiveFromLayout(@Nonnull byte[] data) {
        // Convert from byte data to String
    }

    /** {@inheritDoc} */ 
    @Override
    protected void sendStringToLayout(@Nonnull String value) throws JmriException {
        // Convert string to byte array and then use the method sendToLayout() to send the data.
    }
```

bobjacobsen commented:
I think separate "commanded" and "known" is the right way to go.  I've implemented that in the code I just pushed.  It's also got a working LCC/OpenLCB implementation.

Not so sure about generalizing "String" to "String-in-various-coding" or "byte array".  Java's internal Strings are UTF; no other options.  You can convert on the way in and out of Java to talk to other systems. LCC/OpenLCB, for example, defines strings are to be coded in UTF-8, so that's used for the byte strings that are sent to the hardware.  Other hardware connections are likely to define their own specific character sets.

danielb987 commented:
I see. I thought that OpenLCB/LCC had an undefined charset but since it's utf8, there is no need to select charset. So I will leave that.

dsand47 commented on java/src/jmri/jmrix/openlcb/OlcbStringIOManager.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    @Nonnull
+    public StringIO createNewStringIO(String sName, String uName) {
```
The user name is not being added to the new bean.

bobjacobsen replied:
Fixed.  Thanks.

dsand47 commented:
Various tests are working properly.  I have a script that tests the PropertyChangeEvents.
```
time = 1735363158.58, property = CommandedValue, old = asdf, new = 1234
time = 1735363158.58, property = KnownValue, old = asdf, new = 1234
```

The LCC Monitor captures the events.
```
23:19:18.588: [[195b70b0] 05 01 01 01 88 80 01 00]  S: Unknown message [195b70b0] 05 01 01 01 88 80 01 00
23:19:18.588: [[195b50b0] 31 32 33 34            ]  S: 02.01.12.FE.8E.01                     Producer/Consumer Event Report  EventID:05.01.01.01.88.80.01.00 payload of 4 : 31.32.33.34
   Uses: StringIO SIO Test
```

bobjacobsen commented on help/en/html/hardware/openlcb/index.shtml:
```diff
@@ -52,14 +52,14 @@
       "<code>x0102030405060709</code>".</p>
 
       <p>You can also assign your own names to EventID and use those to create 
```
Fixed, thanks

bobjacobsen commented on help/en/html/hardware/openlcb/index.shtml:
```diff
@@ -153,13 +153,27 @@
+      <p>
+      OpenLCB connections can use StringIOs to create shared variables across
+      multiple JMRI instances.  You create the same StringIO with each JMRI
+      instance, using the same Event ID.  Once you've done that, updating the 
```
Added some text about this.

danielb987 replied:
Thanks!

danielb987 commented:
@dsand47 
This PR looks good to me. Do you have anything to add? Thanks.

bobjacobsen commented:
It's ready from my end.

dsand47 reviewed (approved):
My LCC tests work so it look good to me.
